Sources/SwiftParser/Lexer.swift

```diff
@@ -1448,47 +1497,58 @@ extension Lexer.Cursor {
 
       let tmp = self
       if self.advance(if: { $0.isValidIdentifierContinuationCodePoint }) {
-        diagnosticHandler?(TokStart.distance(to: tmp), InvalidFloatingPointExponentDigit(kind: .digit(Unicode.Scalar(tmp.peek()))))
+        let errorOffset = TokStart.distance(to: tmp)
         self.advance(while: { $0.isValidIdentifierContinuationCodePoint })
-        return (.floatingLiteral, [.isErroneous])
+        return LexerResult(
+          .floatingLiteral,
+          error: .invalidFloatingPointExponentDigit,
+          errorOffset: errorOffset
+        )
       }
     }
 
-    return (.floatingLiteral, [])
+    return LexerResult(.floatingLiteral)
   }
 
-  mutating func lexHexNumber(
-    _ TokStart: Lexer.Cursor,
-    _ diagnosticHandler: ((Int, DiagnosticMessage) -> Void)?
-  ) -> (RawTokenKind, Lexer.Lexeme.Flags) {
+  mutating func lexHexNumber(_ TokStart: Lexer.Cursor) -> LexerResult {
     // We assume we're starting from the 'x' in a '0x...' floating-point literal.
     assert(self.peek() == UInt8(ascii: "x"), "not a hex literal")
     assert(self.previous == UInt8(ascii: "0"), "not a hex literal")
 
     // 0x[0-9a-fA-F][0-9a-fA-F_]*
     _ = self.advance()
-    guard !self.isAtEndOfFile, Unicode.Scalar(self.peek()).isHexDigit else {
-      if !self.isAtEndOfFile {
-        diagnosticHandler?(TokStart.distance(to: self), InvalidDigitInIntegerLiteral(kind: .hex(Unicode.Scalar(self.peek()))))
-      }
+    guard !self.isAtEndOfFile else {
       self.advance(while: { $0.isValidIdentifierContinuationCodePoint })
-      return (.integerLiteral, [.isErroneous])
+      return LexerResult(.integerLiteral)
```

Comment:
This appears to accept the literal `0x`

Reply:
This one is still an issue too

Reply:
Since this is not a regression, I just opened https://github.com/apple/swift-syntax/issues/1214 so we can fix it in a follow-up PR.